gRPC Interceptors

Jared-Prime opened the pull request:
This PR provides an implementation for distributed tracing for Ruby gRPC services and clients.

<img width="1508" alt="screen shot 2018-03-20 at 4 54 19 pm" src="https://user-images.githubusercontent.com/1060000/37684871-6b9a6b64-2c5f-11e8-8f53-8c9d1ace1ce7.png">

In gRPC nomenclature, an "interceptor" is similar to middleware. When running a service or a client, a programmer must declare intended interceptors for use during runtime; as a result, the Datadog `#patch` implementation merely supplies the constants `GRPC::DatadogClientInterceptor` and `GRPC::DatadogServerInterceptor` for use in interceptor registries. Future work may embed tracing in the interceptor call chain itself, making the instrumentation more transparent to the application developer.

The following types of gRPC calls have been implemented and tested:

- synchronous request/reply
- synchronous streaming requests
- synchronous streaming replies

The following types of gRPC have not yet been implemented and/or tested:

- synchronous bidirectional streams
- asynchronous calls of any kind

I've relied upon Datadog's documentation on their platform as well as documentation of the Datadog Ruby gem, Datadog Python package, the Opentracing.io standard, and the existing Rack implementation in this project. Thanks a million to those contributors, as I couldn't have figured out how to do this without their work.

I plan to use my patch going forward and am grateful for any feedback that will help get this code ready for general usage.

Jared-Prime commented:
Looks like there's a Rubocop complaint.

```
Running RuboCop...
Inspecting 190 files
..............................................................................................................................................................................W...............

Offenses:

test/propagation/grpc_propagator_test.rb:30:5: W: Lint/NestedMethodDefinition: Method definitions must not be nested. Use lambda instead.
    def test_extract ...
    ^^^^^^^^^^^^^^^^

190 files inspected, 1 offense detected
RuboCop failed!
```

I'll submit a commit to fix that sometime today.

---

A couple question for the reviewer(s):
1. Do you prefer squashing the commits for the PR or having the full history?
2. I have left the bidirectional support untested in this PR. Do you prefer I complete testing / verification of that message flow on my end, or is it OK to submit a follow-up once ready?

Thanks,
- Jared

delner commented:
Hey @Jared-Prime thanks for submitting this. Pretty excited about this one!

In answer to your questions:

1. We're not too particular about squashing or full history, as long as the history is meaningful.
2. My first thought, in the interest of thoroughness, would be "yes". Although, I'm not yet very familiar with the gRPC world, and what you mean by bidirectional support/testing. Can you elaborate in more detail what this entails, how prominent/important this particular implementation is to gRPC, and maybe any risk factors you anticipate?

delner commented on lib/ddtrace/contrib/grpc/datadog_client_interceptor.rb:
```diff
@@ -0,0 +1,53 @@
+            service: Datadog.configuration[:grpc][:service_name],
+            span_type: 'grpc'
+          }
+          span = ddtracer.trace(name, tracer_options)
```
Question about `name`: looks like this value would be `server` or `client`.  Is there a more meaningful name that can be given? That might add a little context to what it's doing?

Secondary consideration; in many of the ddtrace integrations, we often prefix name with the name of the integration. Perhaps `grpc.X` might be helpful?

delner commented on lib/ddtrace/contrib/grpc/datadog_client_interceptor.rb:
```diff
@@ -0,0 +1,53 @@
+        private
+
+        def ddtracer
+          @ddtracer ||= @options.fetch(:ddtracer, Datadog.tracer)
```
There is a `tracer` option on the gRPC configuration. Should we replace `Datadog.tracer` with `Datadog.configuration[:grpc][:tracer]`?

delner commented on lib/ddtrace/propagation/grpc_propagator.rb:
```diff
@@ -0,0 +1,55 @@
+    include Ext::DistributedTracing
+
+    def self.inject!(context, metadata)
+      metadata[HTTP_HEADER_TRACE_ID] = context.trace_id.to_s
```
I appreciate the idea of being consistent with the other context propagation headers, but would it make sense to create a new set of constants for gRPC? e.g. `GRPC_METADATA_TRACE_ID`/`GRPC_TRACE_ID`? Even if the values are roughly the same or equivalent?

Might be a good idea to avoid coupling with HTTP implementation in case there's some divergence in implementation later.

delner commented on lib/ddtrace/contrib/grpc/datadog_client_interceptor.rb:
```diff
@@ -0,0 +1,53 @@
+            span_type: 'grpc'
+          }
+          span = ddtracer.trace(name, tracer_options)
+          Datadog::GRPCPropagator.inject!(span.context, metadata)
```
This might be a silly question, but in regards to injecting the context info into the metadata...

If I'm understanding the code correctly, `GRPC` is supposed to provide a `metadata` Hash via `request_response(request: nil, call: nil, method: nil, metadata: nil)` options correct?  Is there a case where it doesn't actually provide this metadata Hash?

I ask here because above, `metadata ||= {}` will initialize the nil value to a Hash, but the `inject!` would end up modifying a local Hash variable that'd end up discarded anyways wouldn't it?

Jared-Prime replied:
That's a good question, and I'm not sure about the answer. I'll look it up. My suspicion is that the Hash is always supplied; I'm matching a method signature that I've seen elsewhere. I'll verify what's really needed here.

delner commented on lib/ddtrace/contrib/grpc/datadog_client_interceptor.rb:
```diff
@@ -0,0 +1,53 @@
+          span = ddtracer.trace(name, tracer_options)
+          Datadog::GRPCPropagator.inject!(span.context, metadata)
+
+          yield
```
Since an interceptor is like middleware, does this `yield` equivocate to the actual gRPC call? In which case, could we wrap the `trace` method around it like:

```
ddtracer.trace(name, tracer_options) do |span|
  yield
end
```

The benefit you'd get here is automatic `span.set_error` and `span.finish` functions rendering the `rescue` and `ensure` blocks as moot.

delner commented on lib/ddtrace/contrib/grpc/datadog_server_interceptor.rb:
```diff
@@ -0,0 +1,54 @@
+        private
+
+        def ddtracer
+          @ddtracer ||= @options.fetch(:ddtracer, Datadog.tracer)
```
Just a note that the same general feedback from `DatadogClientInterceptor` likely applies here, too.

delner commented on spec/support/grpc_helpers.rb:
```diff
@@ -0,0 +1,64 @@
+require 'grpc'
+
+def run_service
```
Not a major thing, but would it be more appropriate to put this method into a module? The module could then be composed into the RSpec examples using `include GRPCHelpers`, while having the benefit of not having a global level `run_service` method.

delner commented on test/propagation/grpc_propagator_test.rb:
```diff
@@ -0,0 +1,48 @@
+require 'ddtrace/tracer'
+require 'ddtrace/propagation/grpc_propagator'
+
+class GRPCPropagatorTest < Minitest::Test
```
We've been working on moving away from Minitest towards RSpec. Does this test benefit from something in Minitest that isn't available or conducive to RSpec?

Jared-Prime replied:
No major benefit of Minitest over RSpec. I will rewrite this in RSpec instead

Jared-Prime commented:
Thanks @delner. I've refactored a bit according to your recommendations. I also found that my test for bidirectional (bidi) streaming was simply broken; I've fixed it and able to confirm the code should work for bidi streams as well.

No rush on reviewing; I will be on vacation the rest of this week and next. I'll look forward to tying up any loose ends when I'm back.

delner commented:
General thought when looking at this implementation: would it be conceivable to use GRPC to connect to different services? And if that's the case, would it be reasonable to want to trace those GRPC calls as different services?

If so, we might want to change how we tag traces with `service_name`. I think `configuration[:grpc][:service_name]` is a good start, but would inevitably pool multiple services as one; probably not something most users would want.

To this end we have the `Datadog::Pin` object, which would allow us to tag an individual object with configuration. Typically, we might "pin" service configuration onto each "client" object with `Datadog.configure(client, service_name: 'client_1`)`. Then for the `trace` call, we'd retrieve the configuration from the client via `pin = Datadog::Pin.get_from(client)` and use that to configure the trace.

Does GRPC have a "client" or something similar we could pin configuration on? If not, maybe we should think about some strategy to support GRPC calls as different services.

delner commented:
Also:

 - RSpec examples for GRPC currently aren't running in our CI. If you add a command to run the examples here, they should run. https://github.com/Jared-Prime/dd-trace-rb/blob/57846c44648529378f6caf1147ee9915a09712fd/Rakefile#L233
 - We should rebase this on top of `0.12-dev`. We put new features like this in our current dev branch, so I think this would be a suitable base.

Just to give you a sense of what the plan is for this one, I think we'd ultimately like to target a 0.13.0 release for this PR. This is a really awesome, but sophisticated feature that warrants some additional performance testing among other due diligence measures. Our 0.12.0 release will be rolling soon, and instead of rushing this in, it'd be good to give this one the time and attention it deserves.

We'll be opening a `0.13-dev` branch soon, and I think it could be a good candidate for an early beta version of 0.13.0. Next steps between now and then would be to 1) address any outstanding feedback/issues in this PR and 2) develop a performance test plan (and perhaps a sample app) which we can use to verify that everything is in good shape, then finally 3) merge this PR and ship for a beta release.

delner commented:
Just to give you a sense of what the plan is for this one, I think we'd ultimately like to target a 0.13.0 release for this PR. This is a really awesome, but sophisticated feature that warrants some additional performance testing among other due diligence measures. Our 0.12.0 release will be rolling soon, and instead of rushing this in, it'd be good to give this one the time and attention it deserves.

We'll be opening a `0.13-dev` branch soon, and I think it could be a good candidate for an early beta version of 0.13.0. Next steps between now and then would be to 1) address any outstanding feedback/issues in this PR and 2) develop a performance test plan (and perhaps a sample app) which we can use to verify that everything is in good shape, then finally 3) merge this PR and ship for a beta release.

Jared-Prime commented:
> I think we'd ultimately like to target a 0.13.0 release for this PR. This is a really awesome, but sophisticated feature that warrants some additional performance testing among other due diligence measures.

I agree with that choice of release target and your proposed steps towards shipping 👍 . I'll be able to give this branch more love this week.

Jared-Prime commented:
closing this in favor of improved implementation at #403
